core/src/data_sources/splitter.rs

```diff
@@ -103,13 +105,86 @@ impl Splitter for BaseV0Splitter {
         if chunk.len() > 0 {
             chunks.push(chunk);
         }
+        let mut decoded = Vec::new();
+        let mut remaining : Vec<usize> = Vec::new();
+
+        fn decode_chunk_with_remainder(embedder: &Box<dyn Embedder + Sync + Send>, chunk: Vec<usize>) -> Result<(String, Vec<usize>)> {
+            let mut result : String = String::new();
+            let mut remaining : Vec<usize> = Vec::new();
+
+            let mut end = chunk.len();
+            
+            while end > 0 {
+                let decoded_future = embedder.decode(chunk.clone()[0..end].to_vec());
+                let decoded_result = block_on(decoded_future);
+                if decoded_result.is_err() {
+                    // Do we need a warning log here?
+                    end -= 1;
+                    continue
+                } else {
+                    result = decoded_result.unwrap();
+                    remaining = chunk.clone()[end..].to_vec();
+                    break
+                }
+                
+            }
+
+            Ok((result, remaining))
+        }
 
-        // Decode the chunks in parallel.
-        let mut futures = Vec::new();
         for chunk in chunks {
-            futures.push(embedder.decode(chunk));
+            
+            let mut chunk_with_remaining: Vec<usize> = Vec::new();
+            chunk_with_remaining.append(&mut remaining);
+            chunk_with_remaining.append(&mut chunk.clone());
+
+            let (chunk_decoded, chunk_remainder) = decode_chunk_with_remainder(&embedder, chunk_with_remaining)?;
+            if chunk_decoded.len() > 0 {
+                decoded.push(chunk_decoded);
+            }
+            if chunk_remainder.len() > 0 {
+                // If you have failure on multiple chunks in a row, this won't work :/
+                remaining = chunk_remainder;
+            }
         }
+        if remaining.len() > 0 {
```

Comment:
As discussed this should never happen as the remaining already failed.

If you have a remaining at the end of your algorithm you should error.

That being said the algorithm needs to be refactored a bit to respect `max_chunk_size` so this may pop up again or no depending on your overall approach.